### ISSUE TITLE
feat(back): Estimator Report

### ISSUE THREAD

**Issue by MarieS-WiMLDS** [probabl-ai/skore#834]
### Is your feature request related to a problem? Please describe.

The steps of a ML project is roughly the following (without trying to be fully exhaustive yet):  
- Data exploration
  - EDA
  - cleaning
- Feature Engineering
  - ideation
  - evaluation
  - selection
  - imputer when required
- Model
  - hyperparameters tuning
  - evaluation & selection
  - analysis
    - [feature importance](https://scikit-learn.org/1.5/auto_examples/ensemble/plot_forest_importances.html)
    - metrics
      - accuracy, RMSE, precision... 
      - confusion matrix
    - plots
      - [ROC](https://scikit-learn.org/1.5/modules/generated/sklearn.metrics.RocCurveDisplay.html#sklearn.metrics.RocCurveDisplay.from_estimator) & [PR](https://scikit-learn.org/1.5/modules/generated/sklearn.metrics.PrecisionRecallDisplay.html#sklearn.metrics.PrecisionRecallDisplay.from_estimator) curves for classification
      - QQ-plots or residuals vs fitted values plot for regression
    - calibration
    - robustness
    - [learning curves](https://scikit-learn.org/dev/modules/generated/sklearn.model_selection.learning_curve.html)
- Business Eval
  - threshold choice
  - ROI measure (quite related to above)
- Documentation

We tackled the model evaluation & selection with `CrossValidationReporter`. I would very much like to have something that helps the user to have everything they want to know to evaluate their model for supervised learning.  

### Describe the solution you'd like

My `SupervisedModelAnalyzer` (name suggested) would need to answer the 6 points listed above.  
We have several options when it comes to computing the subparts this object: 
1. everything is given by the user at the beginning. heavy, but ensure consistency of the object
2. the object is like an empty shell, and whenever the user is interested in a subpart, they provide the required information. Light at the beginning, but can turn actually heavy in terms of UX + no insurance that there is consistency.
3. Some parameters are shared, let's ask them at the beginning.  

I suggest going for option 3.  
All the items require:
- fitted estimator
- X_validation 
- y_validation

Addition for each: 
- feature importance
  - option to go with permutation or by impurity in the case of trees (possible to have a default, here i'd rather have permutation)
     - in the case of permutation: scoring to evaluate (possible to have a default)
- learning curves
  - training set & target
  - train sizes (possible to have a default, same as sklearn)
- robustness needs to be defined more exactly, it's not something that exists in scikit learn (maybe in other libs?)

### Describe alternatives you've considered, if relevant

cf options above.  

We could have "flying functions" instead of a main object. But then the additional value compared to scikit-learn is scarce, plus in terms of UX i believe that it's interesting to have all at once. 

### additional context

This new object would have a nice "help" section like CrossValidationReporter.   

2025/01/07:  
The name was changed from Model Analyzer to Estimator Report. 

### Ready Checklist

- [x]  Design specification approved
- [x]  Technical specification approved
- [x]  Critical considerations addressed
- [ ]  (Optional) Planning requirements met

**Comment by tuscland** [comment]
1. Modeling workflow steps question:
  You wrote:  
   > - Model
   >   - hyperparameters tuning
   >   - selection
   >   - analysis
   
   What about this order:
   - Model
     - analysis
     - evaluation
       - hyperparameters tuning
     - selection


2. Isn't `CrossValidationReporter` for model evaluation?

3. Why _Supervised_ `ModelAnalyzer`?

4. I like the modular nature of the `ModelAnalyzer`

**Comment by sylvaincom** [comment]
We had a Notion in September about the global workflow `Structure of a boilerplate`:
https://www.notion.so/probabl/2024-09-13-Exploring-LLMs-to-generate-boilerplates-4e00e9d46c4d494ebf0f68494700215e?pvs=4#f637ea96ba0b4c57ae86975e1fd3439c

**Comment by MarieS-WiMLDS** [comment]
To answer your questions and remarks Camille:  
1. I put the analysis as the last phase because I had in mind the analysis of the chosen model. It's quite intertwined actually because to choose a model, you'll analyze it. 
2. The step I called "selection" actually covers evaluation and then selection. I realize it was not clear and therefore completed. I didn't want to become too precise in this workflow because a true one is not linear nor even a DAG, and I didn't need it. 
3. Supervised because (at least for now) what's inside this object is only useful for supervised models. It makes a longer name though.
4. Nice :D. It's really a strong inspiration from the `CrossValidationReporter`. I think our lib should have one framework of interaction, because for users it would be more complicated to understand otherwise.

**Comment by MarieS-WiMLDS** [comment]
## Suggestion of API

```python
from skore import ModelAnalyzer #could be ModelReport based on issue 950 cited above

est_fitted = est.fit(X_train, y_train)
model_report = ModelAnalyzer.from_fitted_estimator(est_fitted, X_val, y_val)
model_report.metrics.accuracy()
model_report.plots.roc()

```

## What's inside a v1

In the top comment of the epic, I described what we want in a full feature as of today. Let's shortlist some to have a v1:  
- the metrics
- the ROC and PR curves for classification

**Comment by MarieS-WiMLDS** [comment]
It's an option yes.  
We are also interested in building another constructor:  `ModelAnalyzer.from_data(est, X_train, y_train, X_val, y_val)` which is cool because it gives us access to the X_train & y_train. it's nice for some functions such as the learning curves.

**Comment by glemaitre** [comment]
> Could we implement a simple constructor instead of a static factory like we did for CrossValidationReporter

@rouk1 to complement @MarieS-WiMLDS answer, I think that one recurrent recurrent trade-off is to decouple computation vs. adding complexity in the factory (or `__init__`) signature.

For the `ModelReport`, you have

```python
ModelReport.from_fitted_estimator(estimator, X_val, y_val)
ModelReport.from_unfitted_estimator(estimator, X_train, y_train, X_val, y_val)
```

But we could have something similar for the `CrossValidationReport`:

```python
CrossValidationReport.from_cv_results(cv_results, X, y)
CrossValidationReport.from_estimator(estimator, X, y)
```

I did not push towards this API in the `CrossValidationReport` yet because I think that: (i) the user will need to pass the right parameters in the `cross_validate` and data in `from_cv_results` that is more complex, and (ii) the most usual pattern is that the user will actually need to compute the cross-validation.

For the `ModelReport`, it the trade-off is a bit different: it is common for a user to get a fitted model and it is costly to fit a model and you are wasting resources if you already have one fitted. But a fitted model can only provide a partial analysis (since you don't have the training data).

So as a conclusion, I'm personally happier with always the two factories to decouple computation from reporting but I wanted to have user feedback: it is always easier to put a factory back rather than removing it.

**Comment by rouk1** [comment]
In my experience if something takes time it is indeed better to remove it from a constructor and do a two step initialization. 
But two step init has a steeper leraning curve, so ok I understand why you chose the factory pattern. 

I suppose that we should refactor existing stuff to be consistent everywhere.

**Comment by glemaitre** [comment]
As described in #997, there is an alternative API allowing people to not understand what is a class method and go with the usual constructor: introduce a `fit` parameter. It can be `True` and `False but more importantly, the default could be `"auto"`: if the estimator is not fitted then, it try to fit it with the provided training data. If no training data is given, we raise an error.
